vm/src/exceptions.rs

```diff
@@ -526,218 +533,523 @@ pub fn create_exception_type(name: &str, base: &PyTypeRef) -> PyTypeRef {
-        struct $class_name {}
-
-        // We need this to make extend mechanism work:
-        impl PyValue for $class_name {
-            fn class(vm: &VirtualMachine) -> &PyTypeRef {
-                &vm.ctx.exceptions.$ctx_name
-            }
-        }
-
-        #[pyimpl(flags(BASETYPE, HAS_DICT))]
-        impl $class_name {
-            #[pyslot]
-            fn tp_new(
-                cls: PyTypeRef,
-                args: FuncArgs,
-                vm: &VirtualMachine,
-            ) -> PyResult<PyRef<PyBaseException>> {
-                // We need this method, because of how `CPython` copies `init`
-                // from `BaseException` in `SimpleExtendsException` macro.
-                // See: `BaseException_new`
-                PyBaseException::tp_new(cls, args, vm)
-            }
-
-            #[pymethod(magic)]
-            fn init(
-                zelf: PyRef<PyBaseException>,
-                args: FuncArgs,
-                vm: &VirtualMachine,
-            ) -> PyResult<()> {
-                // We need this method, because of how `CPython` copies `init`
-                // from `BaseException` in `SimpleExtendsException` macro.
-                // See: `(initproc)BaseException_init`
-                $base_class::init(zelf, args, vm)
-            }
-        }
+        extend_exception!($exc_struct, $ctx, $class, {});
+    };
+    (
+        $exc_struct:ident,
+        $ctx:expr,
+        $class:expr,
+        { $($name:expr => $value:expr),* $(,)* }
+    ) => {
+        $exc_struct::extend_class($ctx, $class);
+        extend_class!($ctx, $class, {
+            $($name => $value,)*
+        });
     };
 }
 
-// Exception types that extend `BaseException`,
-// sorted the same way CPython does.
+// Sorted By Hierarchy then alphabetized.
 
-extends_exception! {
-    PyException,
+define_exception! {
+    PySystemExit,
     PyBaseException,
-    exception_type,
-    "Common base class for all non-exit exceptions."
+    system_exit,
+    "Request to exit from the interpreter."
 }
-
-extends_exception! {
+define_exception! {
     PyGeneratorExit,
     PyBaseException,
     generator_exit,
     "Request that a generator exit."
 }
-
-extends_exception! {
-    PySystemExit,
-    PyBaseException,
-    system_exit,
-    "Request to exit from the interpreter."
-}
-
-extends_exception! {
+define_exception! {
     PyKeyboardInterrupt,
     PyBaseException,
     keyboard_interrupt,
     "Program interrupted by user."
 }
 
-// Exception types that extend `Exception`,
-// sorted the same way CPython does.
-
-extends_exception! {
-    PyTypeError,
+// Base `Exception` type
+define_exception! {
     PyException,
-    type_error,
-    "Inappropriate argument type."
+    PyBaseException,
+    exception_type,
+    "Common base class for all non-exit exceptions."
 }
 
-extends_exception! {
+define_exception! {
+    PyStopIteration,
+    PyException,
+    stop_iteration,
+    "Signal the end from iterator.__next__()."
+}
```

Comment:
This is a weak suggestion. If this one would be not a maro_rules but a proc-macro like now, I think we can keep it more structural way for readability and future extension.

```rust
/// Signal the end from iterator.__next__().
#[pyexception(name="stop_iteration", base = "PyException")]
struct PyStopIteration {}
```

This code have a little bit more code but understandable like other common `#[pyclass]`es

Reply:
Sure! 👍

Reply:
It is problematic to pass `__init__` and `tp_new` this way. Since proc-macro only allows `key=string` pairs, not `key=ident`
Writting 

```rust
#[pyexception(base = "PyException", ctx_name = "stop_iteration", init="import_error_init")]
struct PyImportError {}

fn import_error_init() { ... }
```

seems a bit off.

Maybe something like:

```rust
#[pyexception(stop_iteration, base_exception_new, import_error_init)]
#[pyclass(base = "PyException")]
struct PyImportError {}
```
?

But, this does not look way better than `define_exception!` macro right now.

Reply:
It is even more verbose, because `pyclass` requires `name=` to be set:

```rust
/// Some doc
#[pyexception(stop_iteration, base_exception_new, import_error_init)]
#[pyclass(name = "ImportError", base = "PyException")]
struct PyImportError {}
```

Reply:
Looks like even this case does not require `struct` fields: https://github.com/RustPython/RustPython/pull/2960/files#diff-d19484f26ba252a3b3ff27867db9682f948572dd37e83e811d6f073bfc2f07c2R644

Reply:
- new and init:`tp_new` and `init` can be added in `pyimpl`.
- `#[pyexception]` can trigger `#[pyclass]` inside the macro. So they shouldn't be stacked.
- I don't think string and ident matter. We already use string literal for pyclass

Reply:
> new and init:tp_new and init can be added in pyimpl.

Do you mean like manually? Or do you mean with some extra macro syntax?

Reply:
I expect finally it looks like:
```rust
#[pyexception(...)]
struct ImportError {}

#[pyimpl]
impl importError {
    #[pyslot]
    fn new(...) {}

    #[pymethod(magic)]
    fn init(...) {}
}
```

I don't think declaring a free function `import_error_init` and use the name in the macro have advantage to the traditional way.